cpp/src/arrow/dataset/file_base.cc

```diff
@@ -313,12 +313,16 @@ Status FileSystemDataset::Write(const FileSystemDatasetWriteOptions& write_optio
         ARROW_ASSIGN_OR_RAISE(auto groups, write_options.partitioning->Partition(batch));
         batch.reset();  // drop to hopefully conserve memory
 
+        if (static_cast<int>(groups.batches.size()) > write_options.max_partitions) {
```

Comment:
Rather, cast `max_partitions` to `int64_t`.

Reply:
will do